pilot/pkg/serviceregistry/kube/controller/endpointcontroller.go

```diff
@@ -102,19 +96,26 @@ func updateEDS(c *Controller, epc kubeEndpointsController, ep interface{}, event
 	} else {
 		endpoints = epc.buildIstioEndpoints(ep, host)
 	}
-	fep := c.collectWorkloadInstanceEndpoints(svc)
-	c.xdsUpdater.EDSUpdate(c.clusterID, string(host), ns, append(endpoints, fep...))
-	// fire instance handles for k8s endpoints only
-	for _, handler := range c.instanceHandlers {
-		for _, ep := range endpoints {
-			si := &model.ServiceInstance{
-				Service:     svc,
-				ServicePort: nil,
-				Endpoint:    ep,
+	if features.EnableK8SServiceSelectWorkloadEntries {
+		c.RLock()
+		istioSvc := c.servicesMap[host]
+		c.RUnlock()
+		if istioSvc == nil {
+			log.Infof("Handle EDS endpoint: service %s/%s has not been populated", svcName, ns)
+			svc, err := c.client.CoreV1().Services(ns).Get(context.TODO(), svcName, metav1.GetOptions{})
```

Comment:
Would this direct call create any issue similar to POD (I know scale is much different here compared to Pod), but still want to check if there are any other alternatives here?

Reply:
We must never, ever do a synchronous call to api-server in this code path again. Please remove this.. it blows up the whole pilot, as we hit ratelimits and get stuck forever

Reply:
We may miss the workload entry

Reply:
Will it sync when the service comes in? I suspect it does, but need to check If not we need a better mechanism. Note that this one doesn't actually work - this is not always caused by a race condition, you can create Endpoints manually without a Service associated, or out of order, in which case this lookup will not work

Reply:
I think you mean `UpdateServiceShards`, which is only called for non k8s registry, not including kubernetes and serviceentry registry here.

Reply:
We may need equivalent of `UpdateServiceShards` that can called when a service comes for K8s to sync up missed endpoints ?